jgrapht-core/src/main/java/org/jgrapht/graph/BaseIntrusiveEdgesSpecifics.java

```diff
@@ -148,11 +148,40 @@ public void setEdgeWeight(E e, double weight)
      */
     public abstract boolean add(E e, V sourceVertex, V targetVertex);
 
+    protected boolean addIntrusiveEdge(E edge, V sourceVertex, V targetVertex, IE e)
+    {
+        if (e.source == null && e.target == null) { // edge not yet in any graph
+            e.source = sourceVertex;
+            e.target = targetVertex;
+
+        } else if (e.source != sourceVertex || e.target != targetVertex) {
+            // Edge already contained in this or another graph but with different touching
+            // edges. Reject the edge to not reset the touching vertices of the edge.
+            // Changing the touching vertices causes major inconsistent behavior.
+            throw new IntrusiveEdgeException(e);
+        }
+        return edgeMap.putIfAbsent(edge, e) == null;
+    }
+
     /**
      * Get the intrusive edge of an edge.
      * 
      * @param e the edge
      * @return the intrusive edge
      */
     protected abstract IE getIntrusiveEdge(E e);
+
+    protected static class IntrusiveEdgeException
```

Comment:
Instead of an inner class, make this a public top-level class.

Reply:
I changed it accordingly.
But I changed the to arguments to source and target vertex because `IntrusiveEdge` is package-private so users cannot use the class and therefore could not call the constructor.
Or should I make the constructor package-private? If the intention is to only allow users to catch the exception and not to create them, a package-private constructor should be suitable.

By the way the `GraphWalk` class also throws an `InvalidGraphWalkException` that is package-private. So if you want to make the exception catchable by users this should be changed too?

Reply:
Public constructor seems fine to me.

Regarding other existing exceptions, we can consider that in a separate sweeping cleanup issue/pull-request (let's put this one to bed first).  In general, making them explicitly catchable seems like the right thing to do, mostly for helping people write tests and debugging code in cases where they have logic errors in their applications.  Since they're not currently public, I think we can make the changes without having to worry about deprecation.

Reply:
Good.

Make writing tests and debugging easy for users is definitely a good point. And I think it is outweighed by the opposing desire to keep to keep the API small.
I can open a separate PR to address it once this one is completed.